webapp/src/components/backstage/playbook_runs/playbook_run/retrospective.tsx

```diff
@@ -1,13 +1,237 @@
-import React from 'react';
+import React, {useRef, useState} from 'react';
 import styled from 'styled-components';
+import {useIntl} from 'react-intl';
+import {DateTime} from 'luxon';
+import debounce from 'debounce';
 
-const Retrospective = () => {
-    return <Container>{'Retrospective'}</Container>;
+import {PlaybookRun, RunMetricData} from 'src/types/playbook_run';
+import {PlaybookWithChecklist} from 'src/types/playbook';
+import {publishRetrospective, updateRetrospective} from 'src/client';
+import {useAllowPlaybookAndRunMetrics, useAllowRetrospectiveAccess} from 'src/hooks';
+import UpgradeBanner from 'src/components/upgrade_banner';
+import {AdminNotificationType} from 'src/constants';
+import {Timestamp} from 'src/webapp_globals';
+import {AnchorLinkTitle, Content, Role} from 'src/components/backstage/playbook_runs/shared';
+import MetricsData from '../playbook_run_backstage/metrics/metrics_data';
+import Report from '../playbook_run_backstage/retrospective/report';
+import ConfirmModalLight from 'src/components/widgets/confirmation_modal_light';
+import {TertiaryButton} from 'src/components/assets/buttons';
+import {PAST_TIME_SPEC} from 'src/components/time_spec';
+
+interface Props {
+    playbookRun: PlaybookRun;
+    playbook: PlaybookWithChecklist | null;
+    onChange: (playbookRun: PlaybookRun) => void;
+    role: Role;
+}
+
+const DEBOUNCE_2_SECS = 2000;
+
+const Retrospective = ({
+    playbookRun,
+    playbook,
+    onChange,
+    role,
+}: Props) => {
+    const allowRetrospectiveAccess = useAllowRetrospectiveAccess();
+    const {formatMessage} = useIntl();
+    const [showConfirmation, setShowConfirmation] = useState(false);
+    const childRef = useRef<any>();
+    const metricsAvailable = useAllowPlaybookAndRunMetrics();
+
+    if (!allowRetrospectiveAccess) {
+        return (
+            <Container>
+                <AnchorLinkTitle
+                    title={formatMessage({defaultMessage: 'Retrospective'})}
+                    id='retrospective'
+                />
+                <BannerWrapper>
+                    <UpgradeBanner
+                        background={<></>}
+                        titleText={formatMessage({defaultMessage: 'Publish retrospective report and access the timeline'})}
+                        helpText={formatMessage({defaultMessage: 'Celebrate success and learn from mistakes with retrospective reports. Filter timeline events for process review, stakeholder engagement, and auditing purposes.'})}
+                        notificationType={AdminNotificationType.RETROSPECTIVE}
+                        verticalAdjustment={0}
+                        vertical={true}
+                    />
+                </BannerWrapper>
+            </Container>
+        );
+    }
+
+    const confirmedPublish = () => {
```

Comment:
optional, more consistent naming

```suggestion
    const onConfirmPublish = () => {
```